### PR TITLE
Feat/jmp 592/enhance policies

### PR DESCRIPTION
**Pull request by healerz**

### CODE REVIEW

**Comment by healerz** [comment]
@dominikzogg rdy :)

File: src/Policy/AndPolicy.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @param NormalizerContextInterface $context
+     * @param object|mixed               $object
+     *
+     * @return bool
+     */
+    public function isCompliant(NormalizerContextInterface $context, $object): bool
+    {
+        foreach ($this->policies as $policy) {
+            if (false === $policy->isCompliant($context, $object)) {
```
Comment (TiMESPLiNTER) [inline src/Policy/AndPolicy.php:33]: Sometimes you use `!` and sometimes `false !==` maybe standardize the way for comparisons.

File: src/Policy/GroupPolicy.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+     *
+     * @return bool
+     */
+    public function isCompliant(NormalizerContextInterface $context, $object): bool
+    {
+        if ([] === $this->groups) {
+            return true;
+        }
+
+        $groups = $context->getAttribute(self::ATTRIBUTE_GROUPS, []);
```
Comment (TiMESPLiNTER) [inline src/Policy/GroupPolicy.php]:
```suggestion
        $contextGroups = $context->getAttribute(self::ATTRIBUTE_GROUPS, []);
```

File: src/Policy/GroupPolicy.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        foreach ($this->groups as $group) {
+            if (in_array($group, $groups, true)) {
```
Comment (TiMESPLiNTER) [inline src/Policy/GroupPolicy.php]:
```suggestion
            if (in_array($group, $contextGroups, true)) {
```

**Review by TiMESPLiNTER — APPROVED** [review]

File: src/Policy/AndPolicy.php (hunk 1 above)
Comment (healerz) [reply src/Policy/AndPolicy.php:33]: done

File: src/Normalizer/Normalizer.php [hunk 4]
```diff
@@ -130,7 +131,11 @@ private function getFieldsByFieldNormalizationMappings(
     ): array {
         $data = [];
         foreach ($normalizationFieldMappings as $normalizationFieldMapping) {
-            if (!$this->isWithinGroup($context, $normalizationFieldMapping)) {
+            if (true !== $this->isCompliant($context, $normalizationFieldMapping, $object)) {
```
Comment (dominikzogg) [inline src/Normalizer/Normalizer.php]: `if (true !== $this->isCompliant($context, $normalizationFieldMapping, $object)) {`

=>

`if (!$this->isCompliant($context, $normalizationFieldMapping, $object)) {`

type checks here drop the readability for no gain

File: src/Normalizer/Normalizer.php [hunk 5]
```diff
@@ -130,7 +131,11 @@ private function getFieldsByFieldNormalizationMappings(
...
+                continue;
+            }
+
+            if (true !== $this->isWithinGroup($context, $normalizationFieldMapping)) {
```
Comment (dominikzogg) [inline src/Normalizer/Normalizer.php]: type checks here drop the readability for no gain

File: src/Normalizer/Normalizer.php [hunk 6]
```diff
@@ -164,7 +169,11 @@ private function getLinksByLinkNormalizationMappings(
     ): array {
         $links = [];
         foreach ($normalizationLinkMappings as $normalizationLinkMapping) {
-            if (!$this->isWithinGroup($context, $normalizationLinkMapping)) {
+            if (true !== $this->isCompliant($context, $normalizationLinkMapping, $object)) {
```
Comment (dominikzogg) [inline src/Normalizer/Normalizer.php]: type checks here drop the readability for no gain

File: src/Normalizer/Normalizer.php [hunk 7]
```diff
@@ -164,7 +169,11 @@ private function getLinksByLinkNormalizationMappings(
...
+                continue;
+            }
+
+            if (true !== $this->isWithinGroup($context, $normalizationLinkMapping)) {
```
Comment (dominikzogg) [inline src/Normalizer/Normalizer.php]: type checks here drop the readability for no gain

File: src/Normalizer/Normalizer.php [hunk 8]
```diff
@@ -188,6 +213,15 @@ private function getLinksByLinkNormalizationMappings(
      */
     private function isWithinGroup(NormalizerContextInterface $context, $mapping): bool
     {
+        @trigger_error(
```
Comment (dominikzogg) [inline src/Normalizer/Normalizer.php:220]: this deprecation is not solveable by the user, should be only triggered if the groups are set

File: src/Policy/CallbackPolicy.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+    /**
+     * @param NormalizerContextInterface $context
+     * @param object|mixed               $object
+     *
+     * @return bool
+     */
+    public function isCompliant(NormalizerContextInterface $context, $object): bool
+    {
+        return ($this->callback)($context, $object);
```
Comment (dominikzogg) [inline src/Policy/CallbackPolicy.php:32]: didn't know that this works

**Review by dominikzogg — COMMENTED** [review]
please add a integration test similar to the group once:

SerializerIntegrationTest.php

**Comment by dominikzogg** [comment]
https://github.com/chubbyphp/chubbyphp-deserialization/pull/29/files

File: src/Normalizer/Normalizer.php (hunk 4 above)
Comment (healerz) [reply src/Normalizer/Normalizer.php]: i got used to it because of our team and the "pre return type" times

Comment (healerz) [reply src/Normalizer/Normalizer.php]: all 5 occurrences changed

File: src/Normalizer/Normalizer.php (hunk 8 above)
Comment (healerz) [reply src/Normalizer/Normalizer.php:220]: move it further down, after the are groups empty check

File: src/Policy/CallbackPolicy.php (hunk 9 above)
Comment (healerz) [reply src/Policy/CallbackPolicy.php:32]: :)

**Comment by dominikzogg** [comment]
@healerz thx

**Pull request merged**